File: postgresql-async/src/test/scala/com/github/mauricio/async/db/postgresql/pool/SingleThreadedAsyncObjectPoolSpec.scala
/*
 * Copyright 2013 Maurício Linhares
 *
 * Maurício Linhares licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */

package com.github.mauricio.async.db.postgresql.pool

import com.github.mauricio.async.db.pool.{SingleThreadedAsyncObjectPool, PoolExhaustedException, PoolConfiguration}
import com.github.mauricio.async.db.postgresql.{DatabaseTestHelper, PostgreSQLConnection}
import java.nio.channels.ClosedChannelException
import java.util.concurrent.TimeUnit
import org.specs2.mutable.Specification
import scala.concurrent.Await
import scala.concurrent.duration._
import scala.language.postfixOps
import com.github.mauricio.async.db.postgresql.exceptions.ConnectionStillRunningQueryException

class SingleThreadedAsyncObjectPoolSpec extends Specification with DatabaseTestHelper {

  "pool" should {

    "give me a valid object when I ask for one" in {

      withPool {
        pool =>
          val connection = get(pool)
          val result = executeTest(connection)
          pool.giveBack(connection)
          result
      }
    }

    "enqueue an action if the pool is full" in {

      withPool({
        pool =>

          val connection = get(pool)
          val promises = List(pool.take, pool.take, pool.take)

          pool.availables.size === 0
          pool.inUse.size === 1
          pool.queued.size === 3

          executeTest(connection)

          pool.giveBack(connection)

          promises.foreach {
            promise =>
              val connection = Await.result(promise, Duration(5, TimeUnit.SECONDS))
              executeTest(connection)
              pool.giveBack(connection)
          }

          pool.availables.size === 1
          pool.inUse.size === 0
          pool.queued.size === 0

      }, 1, 3)

    }

    "exhaust the pool" in {

      withPool({
        pool =>
          1 to 2 foreach {
            _ => pool.take
          }
          await(pool.take) must throwA[PoolExhaustedException]
      }, 1, 1)

    }

    "it shoudl remove idle connections once the time limit has been reached" in {

      withPool({
        pool =>
          val connections = 1 to 5 map {
            _ =>
              val connection = get(pool)
              executeTest(connection)
              connection
          }

          connections.foreach(connection => await(pool.giveBack(connection)))

          pool.availables.size === 5

          Thread.sleep(2000)

          pool.availables.isEmpty must beTrue

      }, validationInterval = 1000)

    }

    "it should validate returned connections before sending them back to the pool" in {

      withPool {
        pool =>
          val connection = get(pool)
          await(connection.disconnect)

          pool.inUse.size === 1

          await(pool.giveBack(connection)) must throwA[ClosedChannelException]

          pool.availables.size === 0
          pool.inUse.size === 0
      }

    }
    
    "it should not accept returned connections that aren't ready for query" in {

      withPool {
        pool =>
          val connection = get(pool)
          connection.sendPreparedStatement("select 1")

          await(pool.giveBack(connection)) must throwA[ConnectionStillRunningQueryException]
          pool.availables.size === 0
          pool.inUse.size === 0
      }

    }

  }

  def withPool[T](
                   fn: (SingleThreadedAsyncObjectPool[PostgreSQLConnection]) => T,
                   maxObjects: Int = 5,
                   maxQueueSize: Int = 5,
                   validationInterval: Long = 3000
                   ): T = {

    val poolConfiguration = new PoolConfiguration(
      maxIdle = 1000,
      maxObjects = maxObjects,
      maxQueueSize = maxQueueSize,
      validationInterval = validationInterval
    )
    val factory = new ConnectionObjectFactory(this.defaultConfiguration)
    val pool = new SingleThreadedAsyncObjectPool[PostgreSQLConnection](factory, poolConfiguration)

    try {
      fn(pool)
    } finally {
      await(pool.close)
    }

  }

  def executeTest(connection: PostgreSQLConnection) = executeQuery(connection, "SELECT 0").rows.get(0)(0) === 0

  def get(pool: SingleThreadedAsyncObjectPool[PostgreSQLConnection]): PostgreSQLConnection = {
    val future = pool.take
    Await.result(future, Duration(5, TimeUnit.SECONDS))
  }

}

File: postgresql-async/src/test/java/com/github/aysnc/sql/db/integration/pool/SingleThreadedAsyncObjectPoolSpec.kt
package com.github.aysnc.sql.db.integration.pool

import com.github.aysnc.sql.db.integration.DatabaseTestHelper
import com.github.aysnc.sql.db.verifyException
import com.github.jasync.sql.db.exceptions.ConnectionNotConnectedException
import com.github.jasync.sql.db.exceptions.ConnectionStillRunningQueryException
import com.github.jasync.sql.db.invoke
import com.github.jasync.sql.db.pool.AsyncObjectPool
import com.github.jasync.sql.db.pool.ConnectionPool
import com.github.jasync.sql.db.pool.PoolExhaustedException
import com.github.jasync.sql.db.postgresql.PostgreSQLConnection
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ExecutionException
import java.util.concurrent.TimeUnit
import org.assertj.core.api.Assertions.assertThat
import org.awaitility.kotlin.await
import org.awaitility.kotlin.matches
import org.awaitility.kotlin.untilCallTo
import org.junit.Test

class SingleThreadedAsyncObjectPoolSpec : DatabaseTestHelper() {

    @Test
    fun `"pool" should "give me a valid object when I ask for one"`() {

        withPool { pool ->
            val connection = get(pool)
            val result = executeTest(connection)
            pool.giveBack(connection)
            result
        }
    }

    @Test
    fun `"pool" should "enqueue an action if the pool is full" `() {

        withPool(1, 3) { pool ->

            val connection = get(pool)
            val promises: List<CompletableFuture<PostgreSQLConnection>> = listOf(pool.take(), pool.take(), pool.take())

            assertThat(pool.idleConnectionsCount).isEqualTo(0)
            assertThat(pool.inUseConnectionsCount).isEqualTo(1)
            assertThat(pool.futuresWaitingForConnectionCount).isLessThanOrEqualTo(3)
            assertThat(pool.idleConnectionsCount).isEqualTo(0)
            assertThat(pool.inUseConnectionsCount).isEqualTo(1)
            assertThat(pool.futuresWaitingForConnectionCount).isLessThanOrEqualTo(3)

            /* pool.take call checkout that call this.mainPool.action,
            so enqueuePromise called in executorService,
            so there is no guaranties that all promises in queue at that moment
             */
            val deadline = TimeUnit.SECONDS.toMillis(5)
            while (pool.futuresWaitingForConnectionCount < 3 || System.currentTimeMillis() < deadline) {
                Thread.sleep(50)
            }

            assertThat(pool.futuresWaitingForConnectionCount).isEqualTo(3)

            executeTest(connection)

            pool.giveBack(connection).get()

            val pools: List<CompletableFuture<AsyncObjectPool<PostgreSQLConnection>>> = promises.map { promise ->
                val con = promise.get(5, TimeUnit.SECONDS)
                executeTest(con)
                pool.giveBack(con)
            }

            pools.map { it.get(5, TimeUnit.SECONDS) }

            await.untilCallTo { pool.idleConnectionsCount } matches { it == 1 }
            assertThat(pool.inUseConnectionsCount).isEqualTo(0)
            assertThat(pool.futuresWaitingForConnectionCount).isEqualTo(0)
        }
    }

    @Test
    fun `"pool" should "exhaust the pool"`() {

        withPool(1, 1) { pool ->
            (1..2).forEach { _ ->
                pool.take()
            }
            verifyException(ExecutionException::class.java, PoolExhaustedException::class.java) {
                awaitFuture(pool.take())
            }
        }
    }

    @Test
    fun `"pool" should "it should remove idle connections once the time limit has been reached" `() {

        withPool(validationInterval = 1000) { pool ->
            val connections = (1..5).map { _ ->
                val connection = get(pool)
                executeTest(connection)
                connection
            }

            connections.forEach { connection -> awaitFuture(pool.giveBack(connection)) }

            await.untilCallTo { pool.idleConnectionsCount } matches { it == 5 }

            Thread.sleep(2000)

            await.untilCallTo { pool.idleConnectionsCount } matches { it == 0 }
        }
    }

    @Test
    fun `"pool" should "it should validate returned connections before sending them back to the pool" `() {

        withPool { pool ->
            val connection = get(pool)
            awaitFuture(connection.disconnect())

            assertThat(pool.inUseConnectionsCount).isEqualTo(1)

            verifyException(ExecutionException::class.java, ConnectionNotConnectedException::class.java) {
                awaitFuture(pool.giveBack(connection))
            }

            assertThat(pool.idleConnectionsCount).isEqualTo(0)
            assertThat(pool.inUseConnectionsCount).isEqualTo(0)
        }
    }

    @Test
    fun `"pool" should "it should not accept returned connections that aren't ready for query" `() {

        withPool { pool ->
            val connection = get(pool)
            connection.sendPreparedStatement("SELECT pg_sleep(3)")

            verifyException(ExecutionException::class.java, ConnectionStillRunningQueryException::class.java) {
                awaitFuture(pool.giveBack(connection))
            }
            await.untilCallTo { pool.idleConnectionsCount } matches { it == 0 }
            await.untilCallTo { pool.inUseConnectionsCount } matches { it == 0 }
        }
    }

    private fun executeTest(connection: PostgreSQLConnection) =
        assertThat(executeQuery(connection, "SELECT 0").rows.get(0)(0)).isEqualTo(0)

    fun get(pool: ConnectionPool<PostgreSQLConnection>): PostgreSQLConnection {
        val future = pool.take()
        return future.get(5, TimeUnit.SECONDS)
    }
}
